### PR TITLE
DC Leaderboard, BattleTV, Rating Chart, Leaderboard

### LINKED ISSUE #5: Battle TV,LeaderBoard,rating graph,DC LeaderBoard
**[Figma Link](https://www.figma.com/file/yO3edbsblDYwj42L8yYLrI/CodeCharacter-2023?node-id=80%3A120&t=GTLykSBvdJnUQFqI-0)**
## Add Battle TV, LeaderBoard and DC LeaderBoard
- Have Tiers as sub-heading for the leaderBoard( not needed for DC-LeaderBoard.).
- Have filters for leaderboard based on tiers( not needed for DC-LeaderBoard.)
- Integrate with backend
- Include Pagination
## Rating Graph
- Integrate rating history with backend
- Build graph using [charts.js](https://www.npmjs.com/package/chart.js?activeTab=readme) based on the rating history of the user.

### PR DESCRIPTION
**Pull request by vvivek1801**
### Completed
- Recreated the Figma 
- Added Tier marking to the leaderboard
- Added DC Leaderboard
- Added Rating Chart

### Fixes required:

- [x]  Filter Button in Leaderboard
- [x]  Outline in BattleTV
- [x]  Rating Chart needs to be connected to backend

### CODE REVIEW

File: src/components/BattleTV/BattleTV.module.css [hunk 1]
```diff
@@ -1,192 +1,189 @@
...
 }
 
-.battlecardHeader .coinused,
-.battlecardHeader .destruction {
-  font-size: 0.8vw;
+.header__title {
+  font-size: 48px;
+  text-align: center;
+  text-transform: uppercase;
+  letter-spacing: 4.5px;
```
Comment (Ram-20062003) [inline src/components/BattleTV/BattleTV.module.css:29]: `letter-spacing` is not needed ig

File: src/components/BattleTV/BattleTV.module.css [hunk 2]
```diff
@@ -1,192 +1,189 @@
...
+.header {
+  max-width: 1000px;
+  margin: 50px auto;
+  border-radius: 12px;
+  justify-content: center;
+  display: flex;
+  flex-wrap: wrap;
+  align-items: center;
+  padding: 10px 25px;
   color: #e8ecfd;
```
Comment (Ram-20062003) [inline src/components/BattleTV/BattleTV.module.css:16]: `#e8ecfd`->`#FFFFF`

File: src/components/BattleTV/BattleTV.module.css [hunk 3]
```diff
@@ -1,192 +1,189 @@
...
-  font-family: Montserrat;
-  font-style: normal;
-  font-weight: 600;
-  color: #ffffff;
-  margin: 20px auto 0 auto;
-  padding: 0 2%;
+.header__icon {
+  max-width: 50px;
+  height: auto;
+  opacity: 0.7;
```
Comment (Ram-20062003) [inline src/components/BattleTV/BattleTV.module.css:22]: Is this used in `BattleTv.tsx`. If no pls remove this

File: src/components/BattleTV/BattleTV.module.css [hunk 4]
```diff
@@ -1,192 +1,189 @@
...
-.battlecardwin {
-  background-color: #00bc6244;
-}
-.battlecardlose {
-  background-color: #bc170044;
+.list {
+  margin: 0 auto;
+  background: transparent;
+  box-shadow: 0 0 4rem 0 rgba(0, 0, 0, 0.1),
+    0 1rem 2rem -1rem rgba(0, 0, 0, 0.3);
```
Comment (Ram-20062003) [inline src/components/BattleTV/BattleTV.module.css]: Box-shadow is not required

File: src/components/BattleTV/BattleTV.module.css [hunk 5]
```diff
@@ -1,192 +1,189 @@
...
-.battlecard {
-  display: flex;
-  flex-wrap: wrap;
-  /* min-width: 720px; */
-  margin: 1rem auto;
-  width: 80%;
-  border-radius: 0.5rem;
-  justify-content: space-between;
+.ranklist {
+  max-width: 1000px;
```
Comment (Ram-20062003) [inline src/components/BattleTV/BattleTV.module.css]: Do not harcode `px`.
```suggestion
  max-width: 50vw;
```
You can use `vw or vh` or `rem` or `em`.

File: src/components/BattleTV/BattleTV.module.css [hunk 6]
```diff
@@ -1,192 +1,189 @@
...
-  /* height: 4rem; */
+.list .item .pic {
+  width: 2rem;
+  height: 2rem;
+  align-items: center;
   border-radius: 50%;
-  background-size: cover;
-  background-position: center;
   box-shadow: 0 0 1rem 0 rgba(0, 0, 0, 0.2),
     0 1rem 1rem -0.5rem rgba(0, 0, 0, 0.3);
```
Comment (Ram-20062003) [inline src/components/BattleTV/BattleTV.module.css]: Box-shadow can be remoed

File: src/components/BattleTV/BattleTV.module.css [hunk 7]
```diff
@@ -1,192 +1,189 @@
...
   align-items: center;
-  border: 2px solid #242a3c;
-  padding: 10px 2%;
-
-  font-family: Montserrat;
-  font-style: normal;
-  font-weight: 600;
-  color: #ffffff;
+  padding-top: 10px;
+  font-family: 'Red Hat Display', sans-serif;
```
Comment (Ram-20062003) [inline src/components/BattleTV/BattleTV.module.css]: Check fonts with figma . it is incorrect

File: src/components/BattleTV/BattleTV.module.css [hunk 8]
```diff
@@ -1,192 +1,189 @@
...
+
+.list .item .score:first-child {
+  border-top-left-radius: 10px;
+  border-bottom-left-radius: 10px;
+}
+
+.list .item .score:last-child {
+  border-bottom-right-radius: 10px;
+  border-top-right-radius: 10px;
 }
```
Comment (Ram-20062003) [inline src/components/BattleTV/BattleTV.module.css]: What does these do exactly? .

File: src/components/BattleTV/BattleTV.module.css [hunk 9]
```diff
@@ -1,192 +1,189 @@
...
-    width: 95%;
-  }
+.button {
+  background-color: white;
+  color: #0a1c2a;
+  border: 1px;
+  padding: 10px;
+  text-transform: uppercase;
+  border-radius: 8px;
+  font-family: monospace, sans-serif;
```
Comment (Ram-20062003) [inline src/components/BattleTV/BattleTV.module.css]: Check font-family with figma

File: src/components/Leaderboard/DailyLeaderboard.tsx [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+  const handlePageClick = (event: { selected: number }) => {
+    const newOffset = (event.selected * itemsPerPage) % items.length;
+    setItemOffset(newOffset);
+  };
+
+  const fetchLeaderboard = () => {
+    setIsLoaded(false);
+    const leaderboardAPI = new DailyChallengesApi(apiConfig);
+    leaderboardAPI
+      .getDailyChallengeLeaderBoard() // The pagination system is so messed up, I can't be bother to fix it, so this hack
```
Comment (Ram-20062003) [inline src/components/Leaderboard/DailyLeaderboard.tsx]: Fix pagination .

File: src/components/Leaderboard/Leaderboard.module.css [hunk 11]
```diff
@@ -46,28 +44,66 @@ html {
...
-  border-radius: 12px;
-  background: radial-gradient(
-    circle,
-    rgba(184, 207, 205, 1) 0%,
-    rgba(133, 157, 217, 1) 100%
-  );
+  background: transparent;
   box-shadow: 0 0 4rem 0 rgba(0, 0, 0, 0.1),
     0 1rem 2rem -1rem rgba(0, 0, 0, 0.3);
   text-align: center;
```
Comment (Ram-20062003) [inline src/components/Leaderboard/Leaderboard.module.css]: Box-shadow is not needed ig

File: src/components/Leaderboard/Leaderboard.module.css [hunk 12]
```diff
@@ -111,18 +172,24 @@ html {
...
 }
 
 .paginationouter {
   text-align: center;
+  flex: 1;
 }
 
 .pagination {
-  font-family: 'Quicksand', sans-serif;
+  font-family: monospace, sans-serif;
```
Comment (Ram-20062003) [inline src/components/Leaderboard/Leaderboard.module.css:195]: check font with figma

File: src/components/Leaderboard/Leaderboard.tsx [hunk 13]
```diff
@@ -32,32 +38,70 @@ function PaginatedItems() {
...
+  const tierMax = 15;
+  const [tierOffest, setTierOffset] = useState(0);
+
   const itemsPerPage = 8;
   const currentUserName = useAppSelector(user).username;
 
   useEffect(() => {
-    fetchLeaderboard();
+    fetchLeaderboard(0, 10000);
+    console.log('I was called');
```
Comment (Ram-20062003) [inline src/components/Leaderboard/Leaderboard.tsx]: remove logs

File: src/components/Leaderboard/Leaderboard.tsx [hunk 14]
```diff
@@ -32,32 +38,70 @@ function PaginatedItems() {
...
+  }
+
+  const fetchLeaderboard = (startIndex: number, endIndex: number) => {
     setIsLoaded(false);
     const leaderboardAPI = new LeaderboardApi(apiConfig);
     leaderboardAPI
       .getLeaderboard(0, 10000) // The pagination system is so messed up, I can't be bother to fix it, so this hack
       .then(response => {
         setItems(response);
+        console.log(items);
```
Comment (Ram-20062003) [inline src/components/Leaderboard/Leaderboard.tsx]: remove logs

File: src/components/Leaderboard/Leaderboard.tsx [hunk 15]
```diff
@@ -32,32 +38,70 @@ function PaginatedItems() {
...
         setIsLoaded(true);
+        setUsableItems(items.slice(startIndex, endIndex));
+        console.log(usableItems);
```
Comment (Ram-20062003) [inline src/components/Leaderboard/Leaderboard.tsx]: remove logs

File: src/components/RatingChart/RatingChart.tsx [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+  CategoryScale,
+  LinearScale,
+  BarElement,
+  Title,
+  Tooltip,
+  Legend,
+} from 'chart.js';
+import { Bar } from 'react-chartjs-2';
+import Loader from '../Loader/Loader';
+import styles from './RatingHistory.module.css';
```
Comment (Ram-20062003) [inline src/components/RatingChart/RatingChart.tsx:13]: This file doesn't even exist. Did u forget to commit?.How did you define your className with `styles` when it did not even exist? :disappointed:

File: src/components/RatingChart/RatingChart.tsx [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        const label = entry.validFrom;
+        labels.push(label);
+        const rating = entry.rating;
+        ratings.push(rating);
+      });
+  });
+
+  function fetchRatingHistory() {
+    setIsLoaded(false);
+    const ratingHistory = userApi.getRatingHistory(currentUser);
```
Comment (Ram-20062003) [inline src/components/RatingChart/RatingChart.tsx:105]: - You are passing `username` here. But [ratingHistory](https://delta.github.io/codecharacter-server-2023/spec/index.html#/user/getRatingHistory) requires `user-id` as `query-parameter`
- You can get `user-id` from `redux-store`

**Review by Ram-20062003 — CHANGES_REQUESTED** [review]
## UI Changes
### LeaderBoard
- [x] Font size should be increased and there should be extra padding for every column
- [x] Background should be `translucent`
- [x] TextColor is not `white`.It is `#AAAAAA`.
- [x] Don't `avatar` and `username`  as `<td></td>` elements. It should be a single element with `avatar` and `username` together.
- [x] Headings can be added with ```font-weight:100```
- [x] Change the order of ` Won Tied Lost` -> `Won Lost Tied`
- [x] Have an uniform design between all components
- [x] Instead of `next` and `previous` have :arrow_forward: and :arrow_backward: 
- [ ] Fix the pagination properly. It is getting all 10000 now, fix it.
- [ ] API calls should be made through redux.

### DC-LeaderBoard
- [x] It should be same as `Leaderboard`. with columns `Rank`,`Username`,`Score`. Username should come along with `avatar`.
- [x]  Add `Pagination` to DC-LeaderBoard.
- [x] Follow same UI as `LeaderBoard`
- [ ] API calls should be made through redux.

### BattleTV
- [x] Font size should be increased and there should be extra padding for every column.
- [x] Background should be `translucent`
- [x] TextColor is not `white`.It is `#AAAAAA`.
- [x] Headings can be added with ```font-weight:100```
- [x] Instead of `next` and `previous` have :arrow_forward: and :arrow_backward: 
- [x] Add `pagination`
- [x] Remove `Box-Shadow`
- [ ] Fix Border issues.

### Rating graph
- [x] Add a route for rating graph
- [x] `Rating.module.css` is not present
- [x] [ratingHistory](https://delta.github.io/codecharacter-server-2023/spec/index.html#/user/getRatingHistory) accepts `user-id` as query parameter not `username`.
### Common Changes
- [x] Remove `box-shadow` wherever it is **not** required
- [x] Use proper font-family. Check in figma regarding `fonts`
- [x] It is better to use `em , rem ,vw or vh` for `width` and `height` than hardcoding it with `px`
- [x] Remove all `console.logs`

(_Please make these changes. If you any doubts, let us know about it_)

File: src/components/Leaderboard/Leaderboard.tsx [hunk 18]
```diff
@@ -102,21 +146,23 @@ function PaginatedItems() {
                     className={styles.matchButton}
                     onClick={() => handleMatchStart()}
                   >
-                    Start match
+                    Start mat border: 1px;ch
```
Comment (Ram-20062003) [inline src/components/Leaderboard/Leaderboard.tsx]: fix this. change this `match`

File: src/components/Leaderboard/DailyLeaderboard.tsx [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+      .then(response => {
+        setItems(response);
+        setIsLoaded(true);
+      })
+      .catch(error => {
+        if (error instanceof ApiError) Toast.error(error.message);
+      });
+  };
```
Comment (Ram-20062003) [inline src/components/Leaderboard/DailyLeaderboard.tsx]: You can add these redux. no need of storing data. just store status of leaderboard like `Loading`,`Success` and `Error`

**Review by Ram-20062003 — CHANGES_REQUESTED** [review]

File: src/components/BattleTV/BattleTV.module.css (hunk 8 above)
Comment (vvivek1801) [reply src/components/BattleTV/BattleTV.module.css]: Gets the corners to be rounded for the row

**Review by Ram-20062003 — APPROVED** [review]

**Review by Ram-20062003 — APPROVED** [review]

**Review by vigneshd332 — APPROVED** [review]

**Pull request merged**